Sources/GRPCCore/Call/Client/Internal/ClientRPCExecutor+HedgingExecutor.swift

```diff
@@ -0,0 +1,499 @@
+            // Stash the response in case we need it later.
+            lastResponse = response
+
+            // Cancel and wait for the processor to finish.
+            processor.cancel()
+            _ = await processor.value
+
+            let shouldSchedule = await existing?.cancel() ?? true
+            guard shouldSchedule else {
+              continue
+            }
+
+            if let delay = delay {
+              let scheduled = HedgingStateMachine.ScheduledAttempt(after: delay) {
+                event.continuation.yield(.start)
+              }
+              state.setNextHedgeDelayTimer(scheduled)
+            } else {
+              event.continuation.yield(.start)
+            }
+
+          case .use(let scheduled):
+            event.continuation.finish()
+            // Cancel the group and the next scheduled attempt.
+            group.cancelAll()
+            // Cancel any other requests.
+            _ = await scheduled?.cancel()
+
+            // Now handle the response.
+            let result = await Result {
+              try await responseHandler(response)
+            }
+
+            // The response has completed, so cancel the processor now.
+            processor.cancel()
+            _ = await processor.value
+
+            return result
+
+          case .none:
+            ()
+          }
+
+        case .timedOut:
+          group.cancelAll()
+          switch state.cancel() {
+          case .cancel(let scheduled):
+            _ = await scheduled.cancel()
+          case .none:
+            ()
+          }
+
+        case .processedRequest(let result):
+          switch result {
+          case .success:
+            ()
+          case .failure:
+            group.cancelAll()
+            switch state.cancel() {
+            case .cancel(let scheduled):
+              _ = await scheduled.cancel()
+            case .none:
+              ()
+            }
+          }
+        }
+      }
+
+      if let lastResponse = lastResponse {
+        return await Result {
+          try await responseHandler(lastResponse)
+        }
+      } else {
+        return .failure(CancellationError())
+      }
+    }
+  }
+}
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+@usableFromInline
+struct HedgingStateMachine {
+  /// The policy used for hedging this request.
+  @usableFromInline
+  let policy: HedgingPolicy
+  /// The current attempt.
+  @usableFromInline
+  var attempt = 1
+  /// The number of inflight requests.
+  @usableFromInline
+  var outstanding = 0
+  /// Retry throttle.
+  @usableFromInline
+  let throttle: RetryThrottle
+  /// The next scheduled attempt.
+  @usableFromInline
+  var nextAttempt: ScheduledAttempt?
+
+  @usableFromInline
+  struct ScheduledAttempt {
+    @usableFromInline
+    let task: Task<Void, Error>
+
+    @inlinable
+    init(after delay: Duration, _ body: @Sendable @escaping () -> Void) {
+      self.task = Task {
```

Comment:
I think we can make this work without an unstructured task. If I understand this correctly we are just using this `Task` here because we need a handle to the different attempts since we have to cancel everything except one (simply speaking). We can models this in a structured pattern where we spawn a child task per request and inside that child task we setup another task group where one child task is doing the request and the other child task is waiting for a cancellation signal. If we want to cancel we can just trigger that signal and the request attempt will be cancelled.
We should be able to even abstract this away and provide an extension on `TaskGroup` where we have a `addTask(_ :()) -> TaskHandle`. In the end that is just setting up the nested task group and returns a continuation like structure.

Reply:
I played around with this idea a bit. You end up having to have an `AsyncStream` to provide the signal in addition to the extra overhead from the task group and extra task.

It works quite well for queuing up the next attempt but not so well for handling the different attempts. 

The reason is that each attempt signals whether it should be processed or discarded. If it's processed we need to continue running its task but cancel all the others. Returning a `TaskHandle` doesn't let us do this directly. We need a way to identify each attempt and linking that back to its handle. We also need to include that ID when we signal when an attempt should be processed and then cancel all handles which don't have that ID. It's doable but to me it felt harder to follow and more error prone.

Reply:
I understand your reasoning here but I am very concerned about using unstructured tasks here. Nothing is holding us accountable that we clean them up and they might be a source of bugs where we miss this. I am especially thinking about long term maintenance of this code. IMO, we should try everything possible to spell constructs in a structured way and bring any ergonomic deficiencies up with the compiler folks. 

It would be very valuable to understand what construct we need from the language to make this pattern easy. Is it enough if `addTask` would return us a `ChildTask` which we could store in a dictionary with an ID and the `ChildTask` allows us to call `.cancel()` and `.value()`?

Reply:
> IMO, we should try everything possible to spell constructs in a structured way and bring any ergonomic deficiencies up with the compiler folks.

I agree, I just don't think returning a task handle is the right approach for racing the RPCs.

> It would be very valuable to understand what construct we need from the language to make this pattern easy. Is it enough if addTask would return us a ChildTask which we could store in a dictionary with an ID and the ChildTask allows us to call .cancel() and .value()?

I think returning a handle which can be cancelled is useful but wouldn't solve all problems here. Being able to get the value is weird though (and also not useful here).

FWIW I think we can make the RPC racing structured by using the broadcast async sequence to signal which attempt gets to survive. This, with some shared state, allows one attempt to effectively cancel the rest.